Android SDK build scripts

marcprux opened the pull request:
Following up on https://github.com/swiftlang/github-workflows/pull/106 and https://github.com/swiftlang/swift/issues/80788, this PR contains a new `swift-docker/swift-ci/sdks/android/` folder with scripts that will build and upload a full stand-alone Swift Android Swift SDK. It is modeled after the structure of the [swift-ci/sdks/static-linux](https://github.com/swiftlang/swift-docker/tree/main/swift-ci/sdks/static-linux) build scripts.

### Running

The top-level `./build` script installs a host toolchain and the Android NDK, and then invokes `scripts/fetch-source.sh` which will fetch tagged sources for libxml2, curl, boringssl, and swift.

It then applies some patches and invokes `scripts/build.sh`, which will build the sources for each of the specified architectures. Finally, it combines the NDK and the newly built SDKs into a single artifactbundle.  

### Specifying Architectures

By default all the supported Android architectures (`aarch64`, `x86_64`, `aarmv7`) will be built, but this can be reduced in order to speed up the build. This can be useful, e.g., as part of a CI that validates a pull request, as building a single architecture takes around 30 minutes on a standard ubuntu-24.04 GitHub runner, whereas building for all the architectures takes over an hour.

To build an artifactbundle for just the `x86_64` architecture, run:

```
TARGET_ARCHS=x86_64 ./build
```

### Installing and validating the SDK

The `.github/workflows/pull_request.yml` workflow will create and upload an installable SDK named something like: `swift-6.1-RELEASE_android-0.1.artifactbundle.tar.gz`. The results of one of the workflow runs can be seen at https://github.com/swift-android-sdk/swift-docker/actions/runs/14603885089

The workflow will also install the SDK locally and use [swift-android-action](https://github.com/marketplace/actions/swift-android-action) to build and test various Swift packages in an Android emulator.

marcprux committed:
Add Android workflow

marcprux committed:
Merge branch 'swiftlang:main' into main

marcprux committed:
Merge branch 'swiftlang:main' into main

marcprux committed:
Merge branch 'swiftlang:main' into main

marcprux committed:
Build Android image (#1)

* Build Android image

* Checkout without ssh

* Retry build if it fails

* Swift 6.1 Release Dockerfiles (#456)

* Change binutils-gold package dependency on Debian 12 to binutils (#457)

* Update installed packages after nightly platform expansion (#458)

* update nightly-6.1 dependencies

* update nightly-main dependencies

* fix ubuntu images

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Fedora 41 Dockerfile (#464)

* Build Android image

* Build Android image

* Swift 6.1 Release Dockerfiles (#456)

* Change binutils-gold package dependency on Debian 12 to binutils (#457)

* Build Android image

* Build Android image

* Build Android image

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Android SDK build

* Swift Androd build

* Swift Androd build

* Swift Androd build

* Swift Androd build

* Swift Androd build

* Swift Androd build

* Swift Androd build

* Swift Androd build

* Swift Androd build

* Swift Androd build

* Swift Androd build

* Swift Androd build

* Swift Androd build

* Swift Androd build

* Swift Androd build

* Swift Androd build

* Swift Androd build

* Swift Android build

* Swift Android build

* Swift Android build

* Swift Android build

---------

Co-authored-by: Mishal Shah <shahmishal@users.noreply.github.com>
Co-authored-by: Chris McGee <87777443+cmcgee1024@users.noreply.github.com>
Co-authored-by: Justice Adams <107649528+justice-adams-apple@users.noreply.github.com>
Co-authored-by: Andrew Sukach <134116196+sookach@users.noreply.github.com>

marcprux committed:
Merge branch 'swiftlang:main' into main

marcprux committed:
Merge branch 'swiftlang:main' into main

marcprux committed:
Swift Android build 6.2 (#2)

* Swift Android build 6.2

* Swift Android build 6.2

* Swift Android build 6.2

* Start splitting NDK out from the rest of the SDK

* Start splitting NDK out from the rest of the SDK

* Swift Android build 6.2

* Swift Android build 6.2

* Swift Android build 6.2

* Swift Android build 6.2

* Swift Android build 6.2

* Swift Android build 6.2

* Swift Android build 6.2

* Swift Android build 6.2

* Swift Android build 6.2

* Swift Android build 6.2

* Swift Android build 6.2

* Swift Android build 6.2

* Swift Android build 6.2

* Swift Android build 6.2

* Swift Android build 6.2

marcprux committed:
Swift Android build 6.2 (#3)

* Swift Android build 6.2

* Swift Android build 6.2

* Add post-install script to SDK bundle

* Add post-install script to SDK bundle

* Add post-install script to SDK bundle

* Add post-install script to SDK bundle

marcprux committed:
Build SDK in Docker container (#4)

* Add static libraries to post-install script

* Add post-install script to SDK bundle

* Add post-install script to SDK bundle

* Update submodules

marcprux committed:
Merge branch 'swiftlang:main' into main

marcprux committed:
Checkout patches repo instead of using a git submodule

marcprux committed:
Update libcurl to 8.13.0

marcprux committed:
Remove resources that we no longer use

marcprux committed:
Update libcurl to 8.13.0

marcprux committed:
Update libxml2 to 2.14.2

marcprux committed:
Build libxml2, libcurl, and boringssl with support for Android 16kb page sizes

marcprux committed:
Add build-script --extra-cmake-options=-DCMAKE_EXTRA_LINK_FLAGS=-Wl,-z,max-page-size=16384

marcprux committed:
Add 16KB page size linker flags to linker flags in swift-toolset.json

marcprux committed:
Add 16KB page size linker flags to linker flags in swift-toolset.json

marcprux committed:
Build with ndk-r28b

marcprux committed:
Revert to building with ndk-r27c

marcprux committed:
Use official endpoints for discovering latest Swift release/devel/trunk tags

marcprux committed:
Typo fix in version script

marcprux committed:
Cleanup for PR

marcprux committed:
Change BUILD_VERSION to BUILD_SCHEME and have it match release, swift-*-branch, or development

marcprux committed:
Update Android README

marcprux committed:
Update how patches are applied

marcprux committed:
Fix source directory for patch target

marcprux committed:
Harmonize timestamps in artifactbundle with the swift source tag date for build reproducibility

marcprux committed:
Simplify toolchain-vars.sh

marcprux committed:
Merge branch 'swiftlang:main' into main

marcprux committed:
Merge branch 'swiftlang:main' into main

marcprux committed:
Merge branch 'swiftlang:main' into main

marcprux committed:
Merge branch 'swiftlang:main' into main

marcprux committed:
Run the compiler validation suite for Android (#8)

* Run the compiler validation suite for Android

* Add --build-compiler option

* Add --cross-compile-build-swift-tools=0 from https://github.com/swiftlang/swift/pull/38441

* Build with --build-llvm=0

* Check out Yams for Swift 6.1.1 build

* Add --llvm-ninja-targets-for-cross-compile-hosts=help

* Install pre-requisites

* Install build prerequisites

* Only setup local toolchain if build-compiler is not 0

* Fix Yams version checkout

* Quote arguments to build scripts

* Permit empty host-toolchain argument in build.sh

* Fix check for BUILD_COMPILER

* Re-order Docker PATH to system clang is used before Swift toolchain clang

* Use --host-test to skip attempt to test on connected device/emulator

* Install clang in Dockerfile

* Add --skip-test-linux flag to build

* Remove more folders to free up space

* Update patches

* Add docker-specific CI variants that run the compiler validation tests

* meaningless edit

* Update pull_request.yml to stop the 6.1 release builds

* Update pull_request.yml to really only build the full docker compiler with tests

* Update README.md with meaningless edit to bump build

* Update build.sh to skip testing XCTest for linux, as a handful of the linux tests fail for some reason

* Update build.sh to disable building libTesting and for 16K memory pages, as both don't work yet

* Update build.sh to skip testing Foundation for linux, which requires building SwiftPM from source first

* Update build.sh to not build each arch in a separate build directory

* Update Dockerfile to use clang 19 instead

* Add self-hosted runner CI variant

* Fix CI workflow syntax

* Update CI runner config

* Update CI runner config

* Fix name of self-hosted CI runner

* Re-order arch run sequence

* Update CI for self-hosted runner

* Build for self-hosted runner

* Add swift-6.2-branch to self-hosted run matrix

* Update build.sh to disable aarch64 temporarily, as we know it built fine

* Update pull_request.yml to disable non-compiler builds, as they all work, and comment out self-hosted till we know armv7 works

* Run compiler validation on self-hosted with increased timeout

* Update build-docker to only build for armv7

* Update pull_request.yml to disable self-hosted builds temporarily

* Update pull_request.yml to build the full compiler in github runners with the armv7 stdlib and tests

* Update build-docker to build for aarch64 first

* Update toolchain-vars.sh to use older trunk snapahot toolchain that didn't crash

* Update toolchain-vars.sh to only download the latest release compiler if building the Swift compiler from source

* Update build-docker to only build for armv7 again

* Update toolchain-vars.sh to fix setting branch variable

* Update build-docker to build for all three supported arches

* Update pull_request.yml to try self-hosted runs again

* Update build.sh to only install SDK components and remove linux stdlib

* Tolerate missing linux folder when attempting to clean up unnecessary build artifacts

* Fix extraCLIOptions in swift-toolset.json

* gcpdw

* Fix patch application

* Update apply-patches.sh to remove unused changes

* Update swift-android-testing-except-release.patch to add Testing fix

* Update swift-android.patch to remove unneeded patches

* Update build.sh to try and fix Testing and clean up the bundle more

* Fix README.md

---------

Co-authored-by: finagolfin <finagolfin@tuta.io>

finagolfin committed:
Build compiler-validated bundles from latest branch commits, not older tags

Also update patches, particularly to disable failing tests from the compiler validation suite

finagolfin committed:
Merge pull request #9 from swift-android-sdk/Testing

Update patches and build the compiler from the latest commits, not an older tag

finagolfin committed:
Centralize cmake variable for 6.2 in patches, as done for trunk upstream

finagolfin committed:
Update disabled tests

marcprux committed:
gcpd 'Update patches'

marcprux committed:
Update patches and build matrix

marcprux committed:
Disable compiler validated builds on self-hosted

marcprux committed:
Merge branch 'main' into update-patches2

marcprux committed:
Try running Docker build on macOS host

etcwilde commented on swift-ci/sdks/android/scripts/fetch-source.sh:
```diff
@@ -0,0 +1,194 @@
+}
+
+# Defaults
+if [[ -z "${SWIFT_VERSION}" ]]; then
```
Why are we downloading these separately? `update-checkout` is already pulling down yams, curl, and libxml2. Do we need to update the versions there for Android? Is this just to keep those versions clean while applying patches to the versions downloaded by this script?

marcprux replied:
I didn't realize `update-checkout` grabbed these too. I was just following the precedent of the static-linux SDK here:

https://github.com/swift-android-sdk/swift-docker/blob/ed09034cd0d61b32cbacd56b26a38fa8b5e823fe/swift-ci/sdks/static-linux/scripts/fetch-source.sh#L183-L194

I could likely re-use the existing yams and libxml2 checkouts, and probably curl too, but we are building and linking BoringSSL (like the static linux SDK does), so we will need to make sure that works with the version of curl that is being grabbed with `update-checkout`.

marcprux replied:
I've removed the redundant source checkouts in https://github.com/swift-android-sdk/swift-docker/pull/11/. Thanks for spotting it! @al45tair, do you know why the static-linux SDK is checking out libxml2 and curl when it could be re-using the repositories already checked out with `update-checkout`?

al45tair replied:
I think I wanted to have direct control over which versions get built into the Static SDK (which also let me pick new enough versions that they'd work with the other dependencies).

BTW, I'd be careful with BoringSSL. OpenSSL would be a safer choice, I think, simply because BoringSSL doesn't really have releases and so unless you collaborate directly with the BoringSSL maintainers (which is what NIO does), you won't get security fixes and so on.  I'm actually thinking of switching the Static SDK over to OpenSSL for that reason — either that, or making it always grab the version NIO is currently using, since the NIO folks already watch for updates to BoringSSL.

al45tair replied:
(I wish the BoringSSL maintainers would make it slightly more explicit that using it was a bad idea, rather than merely saying they don't recommend it because of API/ABI stability. The thing I worry about is security bugs, not ABI stability.)

marcprux replied:
I had assumed/presumed that since the BoringSSL "version" being used was the "fips-20220613" tag, that that FIPS designation held some significance or potential for some security-related certification.

The other reason I stuck with it is that BoringSSL is what Android itself uses by default as of Android 6.0/API 23, even though they don't expose it as part of the NDK. I'm sure OpenSSL would be fine too, but I've read claims that BoringSSL is "optimized for Android".

> making it always grab the version NIO is currently using, since the NIO folks already watch for updates to BoringSSL.

I think this would make the most sense. I wonder if we could harmonize on some source of truth for the version, and perhaps `update-checkout` could start checking out BoringSSL with this tag.

Is FoundationNetwork going to be based on swift-nio/swift-nio-ssl in the future? If so, then the issue will solve itself eventually :smile:

Perhaps @Lukasa has insights that would apply here?

al45tair replied:
> I had assumed/presumed that since the BoringSSL "version" being used was the "fips-20220613" tag, that that FIPS designation held some significance or potential for some security-related certification.

Yeah, that was my assumption too, but that's (sadly) wrong. @Lukasa, WDYT we should do here?

finagolfin replied:
> I could likely re-use the existing yams and libxml2 checkouts, and probably curl too, but we are building and linking BoringSSL (like the static linux SDK does), so we will need to make sure that works with the version of curl that is being grabbed with update-checkout.

Note that the `build-script` curl build also links against OpenSSL, but simply does so implicitly, so it only works when natively compiling and the system package is installed, then fails when cross-compiling, so we'll definitely need to add one of those SSL libraries to the `build-script` config.

> I wonder if we could harmonize on some source of truth for the version, and perhaps update-checkout could start checking out BoringSSL with this tag.

`update-checkout` already checks out swift-crypto, which is based on BoringSSL for most platforms and has a CMake config, so maybe we can just reuse that for curl too?

etcwilde replied:
The Windows runtime builds Curl without SSL support: https://github.com/swiftlang/swift/blob/663ec9340d37b3407c8408b0f822d8d9962162b1/utils/build.ps1#L2351

@compnerd, what is the impact on FoundationNetworking and how does it deal with encryption on Windows?

If it's using some other mechanism and we don't actually need it, then maybe we can turn off the SSL support in curl entirely?

marcprux replied:
From the last Android workgroup meeting, we discussed this and decided to stick with the same BoringSSL version that the static Linux SDK is using. In the future, both SDKs might harmonize on switching to OpenSSL together.

etcwilde commented on .github/workflows/pull_request.yml:
```diff
@@ -21,3 +21,172 @@ jobs:
           path: |
             *.log
+
+   android-build:
```
I'm concerned about running this on every PR against the swift-docker repository.
At the moment, testing on this repo takes a few minutes maximum, while I suspect that this will increase that it pretty substantially. It might be good to either limit running it to changes to the android subdirectory, or we should make this a dedicated job on Swift CI like the static Linux Musl SDK instead.
CC @shahmishal, @justice-adams-apple

marcprux replied:
> I'm concerned about running this on every PR against the swift-docker repository.

Yeah, actually, that `android-build` CI part was intended to be expunged before this PR would be merged. 😊 

It is just there so we can exercise updates to the pull request.

etcwilde replied:
Okay, that sounds good to me. Either Justice or I can probably use it as a template for setting up the Jenkins job if we have questions later, so useful to have in history.

etcwilde reviewed (commented):
Started going through this. I think my main concerns right now are around cloning repo's that we are already cloning with update-checkout and the patches that haven't been merged into the repos yet. Otherwise it's generally looking fine. I still need to look more closely at the `build.sh` though.

marcprux committed:
Remove separate checks of libxml2, curl, and yams

marcprux committed:
Merge pull request #11 from swift-android-sdk/update-patches2

Remove unnecessary source checkouts

finagolfin committed:
Remove upstreamed 6.2 branch patches

marcprux committed:
Merge branch 'swiftlang:main' into main

finagolfin committed:
Use command-line flags to work around CMake 3.30+ linker flag bug, rather than patching files

finagolfin committed:
Create an Android CMake toolchain file instead to cross-compile Testing and add
16 KB page linker flag to 6.2 branch

finagolfin committed:
Install the native host LLVM tools for full compiler builds

finagolfin committed:
Extend `--cross-compile-build-swift-tools=False` to disable building and installing LLVM
and the Testing macros for cross-compiled hosts, then enable building and using the
Testing macros for the native host.

etcwilde commented on swift-ci/sdks/android/scripts/install-swift.sh:
```diff
@@ -0,0 +1,54 @@
+chmod -R o+r ${destination}/lib/swift
+
+popd >/dev/null
+
```
Nit: extra newline

marcprux replied:
Fixed in https://github.com/swift-android-sdk/swift-docker/pull/14

etcwilde commented on swift-ci/sdks/android/scripts/fetch-source.sh:
```diff
@@ -0,0 +1,152 @@
+git checkout ${BORINGSSL_VERSION}
+popd >/dev/null 2>&1
+groupend
+
```
Nit: extra newline

marcprux replied:
Fixed in https://github.com/swift-android-sdk/swift-docker/pull/14

etcwilde commented on swift-ci/sdks/android/README.md:
```diff
@@ -0,0 +1,85 @@
+the same artifacts in approximately half the time, and
+may be suitable to an already containerized envrionment (such as
+a GitHub runner).
+
```
Do we want to include an example of the command(s) for running the build locally?

marcprux replied:
Added in https://github.com/swift-android-sdk/swift-docker/pull/14

etcwilde commented on swift-ci/sdks/android/scripts/build.sh:
```diff
@@ -0,0 +1,704 @@
+            -DBUILD_STATIC_LIBS=ON \
+            -DCURL_BUILD_TESTS=OFF
+
+        quiet_pushd ${build_dir}/$arch/curl
```
Alternatively, save the pushd/popd call and let CMake run the build for you.

```sh
run cmake --build ${build_dir}/$arch/curl -- -j ${parallel_jobs}
```

marcprux replied:
Fixed in https://github.com/swift-android-sdk/swift-docker/pull/14

etcwilde commented on swift-ci/sdks/android/scripts/build.sh:
```diff
@@ -0,0 +1,704 @@
+
+        header "Installing libcurl for $arch"
+        quiet_pushd ${build_dir}/$arch/curl
+            run ninja -j$parallel_jobs install
```
```sh
run cmake --install ${build_dir}/${arch}/curl
```

marcprux replied:
Fixed in https://github.com/swift-android-sdk/swift-docker/pull/14

etcwilde commented on swift-ci/sdks/android/scripts/build.sh:
```diff
@@ -0,0 +1,704 @@
+    groupend
+
+    groupstart "Building libcurl for ${compiler_target_host}"
+    quiet_pushd ${swift_source_dir}/curl
```
We're already passing `-S` and `-B` to CMake. We shouldn't need the `pushd/popd`.

marcprux replied:
Fixed in https://github.com/swift-android-sdk/swift-docker/pull/14

etcwilde commented on swift-ci/sdks/android/scripts/build.sh:
```diff
@@ -0,0 +1,704 @@
+
+# Copy necessary headers and libraries from the toolchain and NDK clang resource directories
+mkdir -p $swift_res_root/usr/lib/swift/clang/lib
+cp -r $host_toolchain/lib/clang/*/include $swift_res_root/usr/lib/swift/clang
```
The clang headers should be found in the compiler resource-dir, not the SDK. If things get out of sync, this can explode violently.

finagolfin replied:
I just deleted this altogether in the latest commit, having the post-install setup script link the bundle to the NDK's clang resource directory instead. That worked fine in my testing.

etcwilde commented on swift-ci/sdks/android/scripts/build.sh:
```diff
@@ -0,0 +1,704 @@
+    declare -g ${name}_license="$license"
+    declare -g ${name}_url="$url"
+
+    packages+=(${name})
```
We should use this to generate the spdx SBOM for the SDK like the static SDK does here: https://github.com/swiftlang/swift-docker/blob/3bf68fbcd344e968dfda10e5aa236978283c9310/swift-ci/sdks/static-linux/scripts/build.sh#L757-L871

marcprux replied:
OK, I've added generation of a `sbom.spdx.json` file in https://github.com/swift-android-sdk/swift-docker/pull/14, which will look like:

```json
{
  "SPDXID": "SPDXRef-DOCUMENT",
  "name": "SBOM-SPDX-9fe120c9-d845-4586-876b-4080c9143143",
  "spdxVersion": "SPDX-2.3",
  "creationInfo": {
    "created": "2025-08-22T02:37:12+00:00",
    "creators": [
      "Organization: Apple, Inc."
    ]
  },
  "dataLicense": "Apache-2.0",
  "documentNamespace": "urn:uuid:cceb1cb9-34b6-4595-9e1f-9091c408d88b",
  "documentDescribes": [
    "SPDXRef-Package-swift-android-sdk"
  ],
  "packages": [
    {
      "SPDXID": "SPDXRef-Package-swift-android-sdk",
      "name": "Swift SDK for Android",
      "versionInfo": "",
      "filesAnalyzed": false,
      "licenseDeclared": "Apache-2.0",
      "licenseConcluded": "Apache-2.0",
      "downloadLocation": "https://swift.org/install",
      "copyrightText": "NOASSERTION",
      "checksums": []
    },
    {
      "SPDXID": "SPDXRef-Package-swift",
      "name": "swift",
      "versionInfo": "6.2-DEVELOPMENT-SNAPSHOT-2025-08-20-a",
      "filesAnalyzed": false,
      "licenseDeclared": "Apache-2.0",
      "licenseConcluded": "Apache-2.0",
      "downloadLocation": "https://swift.org",
      "copyrightText": "NOASSERTION",
      "checksums": []
    },
    {
      "SPDXID": "SPDXRef-Package-libxml2",
      "name": "libxml2",
      "versionInfo": "2.11.5",
      "filesAnalyzed": false,
      "licenseDeclared": "MIT",
      "licenseConcluded": "MIT",
      "downloadLocation": "https://github.com/GNOME/libxml2",
      "copyrightText": "NOASSERTION",
      "checksums": []
    },
    {
      "SPDXID": "SPDXRef-Package-curl",
      "name": "curl",
      "versionInfo": "8.9.1",
      "filesAnalyzed": false,
      "licenseDeclared": "MIT",
      "licenseConcluded": "MIT",
      "downloadLocation": "https://curl.se",
      "copyrightText": "NOASSERTION",
      "checksums": []
    },
    {
      "SPDXID": "SPDXRef-Package-boringssl",
      "name": "boringssl",
      "versionInfo": "fips-20220613",
      "filesAnalyzed": false,
      "licenseDeclared": "OpenSSL AND ISC AND MIT",
      "licenseConcluded": "OpenSSL AND ISC AND MIT",
      "downloadLocation": "https://boringssl.googlesource.com/boringssl/",
      "copyrightText": "NOASSERTION",
      "checksums": []
    }
  ],
  "relationships": [
    {
      "spdxElementId": "SPDXRef-Package-swift-android-sdk",
      "relationshipType": "GENERATED_FROM",
      "relatedSpdxElement": "SPDXRef-Package-swift"
    },
    {
      "spdxElementId": "SPDXRef-Package-swift-android-sdk",
      "relationshipType": "GENERATED_FROM",
      "relatedSpdxElement": "SPDXRef-Package-libxml2"
    },
    {
      "spdxElementId": "SPDXRef-Package-swift-android-sdk",
      "relationshipType": "GENERATED_FROM",
      "relatedSpdxElement": "SPDXRef-Package-curl"
    },
    {
      "spdxElementId": "SPDXRef-Package-swift-android-sdk",
      "relationshipType": "GENERATED_FROM",
      "relatedSpdxElement": "SPDXRef-Package-boringssl"
    }
  ]
}
```

etcwilde commented on swift-ci/sdks/android/scripts/build.sh:
```diff
@@ -0,0 +1,704 @@
+ndk_sysroot="ndk-sysroot"
+
+cat > scripts/setup-android-sdk.sh <<'EOF'
+#/bin/bash
```
Should this be a shebang?

```suggestion
#!/bin/env bash
```

marcprux replied:
I've switched it to be #!/usr/bin/env bash

etcwilde reviewed (commented):
Noticed an issue in one of the shebangs.
Do we want a `/usr/bin/env bash`? I know these scripts are documented to work on Ubuntu 24.04 so the `/bin/bash` should be fine, but it looks like most of the pieces should work on other linux distros too.

finagolfin committed:
Don't copy any libraries from the NDK: link the NDK's clang resource directory in the post-install script instead

finagolfin committed:
Remove upstreamed patches and turn the last modifications into perl substitutions instead.

marcprux committed:
Wildcard the clang version link in the post-install script in order to accomodate NDK 27 and 28

marcprux committed:
Merge pull request #13 from swift-android-sdk/clang-wildcard

Wildcard the clang version link in the post-install script in order to accomodate NDK 27 and 28

marcprux committed:
Merge branch 'swiftlang:main' into main

marcprux committed:
Switch generated shell header to use #!/usr/bin/env bash

marcprux committed:
SBOM and nits (#14)

* Address nits

* Eliminate unnecessaryt pushds and use run cmake install instead of ninja

* Generate SBOM with SDK dependencies

* Add example of command for building locally

finagolfin committed:
Remove self-hosted build on github CI

etcwilde reviewed (approved):
I think this looks good. Thank you for working through it.

shahmishal commented on .github/workflows/pull_request.yml:
```diff
@@ -21,3 +21,155 @@ jobs:
+          echo 'import Dispatch' >> Sources/DemoProject/DemoProject.swift
+          echo 'import Android' >> Sources/DemoProject/DemoProject.swift
+      - name: Test Demo Project on Android
+        uses: skiptools/swift-android-action@main
```
What are we using this action for?

finagolfin replied:
Nothing, this entire github workflow change is just a demo. We're looking for guidance on how you want to do it.

marcprux replied:
As we [discussed](https://github.com/swiftlang/swift-docker/pull/467/files#diff-a0fe23534b616d51ce686d2a1bcd1a78bc75074aef1a2f6ee96c9469991e1a4cR25) in the review for the file, that whole `android-build` section of the workflow is just for validating the build process while we are iterating on the PR. It has been very useful for ensuring that all the Android SDK build variants run successfully, but we figured we might not want it in the final PR workflow.

I've excised that whole section in [8eb982e](https://github.com/swiftlang/swift-docker/pull/467/commits/8eb982e88729c91a3cc69760d6c32d78d93deaa0).

shahmishal commented on .github/workflows/pull_request.yml:
```diff
@@ -21,3 +21,155 @@ jobs:
+          repository: apple/swift-algorithms
+          path: swift-algorithms
+      - name: Test swift-algorithms
+        uses: skiptools/swift-android-action@main
```
What are we using this action for?

marcprux replied:
Also removed in [8eb982e](https://github.com/swiftlang/swift-docker/pull/467/commits/8eb982e88729c91a3cc69760d6c32d78d93deaa0)

shahmishal commented on swift-ci/sdks/android/scripts/build.sh:
```diff
@@ -0,0 +1,785 @@
+        # use an out-of-tree build folder
+        export SWIFT_BUILD_ROOT=${build_dir}/swift-project
+
+        ./swift/utils/build-script \
```
Not a blocker: We should create a preset in the swift repo, incase someone changes the flags we don't break android sdk.

finagolfin replied:
My goal is to eventually obsolete this build script by replacing it entirely with a build preset and a call to the `swift-sdk-generator` tool. However, since we couldn't get that done before the 6.2 release, we chose to instead reuse the static linux SDK build script as much as possible, which you can see by just running `diff -ruw` between this new Android build and the current static linux SDK build.

The only differences to the static linux SDK build are that we use the official `build-script` to build much more of the toolchain, had to add a post-install script because configuring external C/C++ platform SDKs like the Android NDK is still partially broken in SwiftPM, swiftlang/swift-package-manager#8958, (the static linux SDK doesn't face this problem because it ships Musl libc and libc++ _in_ the bundle, which we didn't want to do with the C/C++ libraries from the Android NDK), and instead of the single `build` script, we ship both a `build-docker` and a `build-local`, depending on whether you want to use Docker or not (we know you only use Docker on this official CI, but we provided it for the convenience of others who don't want Docker).

marcprux committed:
Remove android_build job from pull_request workflow

marcprux committed:
Add license headers to Android build scripts

marcprux committed:
Remove unnecessary .gitignore

marcprux committed:
Add license header to Dockerfile

marcprux committed:
Merge branch 'swiftlang:main' into main

finagolfin committed:
Switch over to using explicit tags and branches when invoking build-docker/local

Also, force the full Swift compiler to be built from source when building from a
branch, since there are no existing toolchain builds of arbitrary commits that
we can download.

marcprux committed:
Merge branch 'swiftlang:main' into main
